[Module Aliasing] Modify module loaders to use module real names

elsh opened the pull request:
Modify module loaders to use module 'real name' (physical name on-disk) when loading, since it can be different from 'name' if module aliasing is used. Also use the 'real name' to add/retrieve loaded modules in ASTContext. Resolves rdar://83591943.

elsh committed:
Load modules using real names if module aliases are used
rdar://83591943

elsh commented:
@swift-ci smoke test

xymus commented on test/Frontend/load-module-with-alias.swift:
```diff
@@ -0,0 +1,20 @@
+// RUN: %empty-directory(%t)
+
+/// Create a module Bar
+// RUN: echo 'public func bar() {}' > %t/FileBar.swift
```
Just FYI as this is a personal preference kind of thing, you may gain to use a `RUN: %{python} %utils/split_file.py -o %t %s` to break apart the test files into files ready to be compiled instead of echoing them like this.

xymus commented on lib/Sema/SourceLoader.cpp:
```diff
@@ -34,13 +34,20 @@ using namespace swift;
-static FileOrError findModule(ASTContext &ctx, StringRef moduleID,
+static FileOrError findModule(ASTContext &ctx, Identifier moduleID,
                               SourceLoc importLoc) {
   llvm::SmallString<128> inputFilename;
```
When is the source loader used? Should we test this?

elsh replied:
done

xymus reviewed (approved):
This looks good!

CodaFi commented:
I assume the frontend is meant to accept multiple aliases for the same module, and that they'll stomp over each other such that we'll only take the last one? E.g. `-module-alias Foo=Bar -module-alias Foo=Baz` means `Foo=Baz` wins? Could you test that please?

CodaFi commented:
Are module aliases intended to propagate through swift interface files as well?

davidungar reviewed (commented):
No opinion. Not anything I know about, sorry.

elsh commented:
> I assume the frontend is meant to accept multiple aliases for the same module, and that they'll stomp over each other such that we'll only take the last one? E.g. `-module-alias Foo=Bar -module-alias Foo=Baz` means `Foo=Baz` wins? Could you test that please?

Only one alias per module is allowed; the input validation is done at https://github.com/apple/swift/pull/39376. It takes in multiple -module-alias inputs where each one maps to a unique alias, and if there's a duplicate, it will throw an error.

hamishknight commented on test/Frontend/load-module-with-alias.swift:
```diff
@@ -0,0 +1,20 @@
+// RUN: test -f %t/Bar.swiftmodule
+// RUN: not test -f %t/Cat.swiftmodule
+
+/// Create a module Foo that imports Cat with -module-alias Cat=Bar
+// RUN: echo 'import Cat' > %t/FileFoo.swift
+// RUN: %target-swift-frontend -module-name Foo -module-alias Cat=Bar %t/FileFoo.swift -emit-module -emit-module-path %t/Foo.swiftmodule -I %t
```
Would it also be worth having a test for when the module "Cat" does exist on disk, but the module alias means another module gets loaded?

elsh replied:
yup will add the test

hamishknight reviewed (commented):
Looks good to me! @CodaFi raises a good point regarding swift module interfaces, as I understand it, it could be handled by either:

1. Writing the module alias information into the interface files and serializing it into the swiftmodules, such that clients can recover the real module names and load the right dependencies
2. Canonicalizing the modules names to the real module names when writing out interfaces and serializing modules

elsh commented:
> Looks good to me! @CodaFi raises a good point regarding swift module interfaces, as I understand it, it could be handled by either:
> 
> 1. Writing the module alias information into the interface files and serializing it into the swiftmodules, such that clients can recover the real module names and load the right dependencies
> 2. Canonicalizing the modules names to the real module names when writing out interfaces and serializing modules

#1 is the plan. Will add a test.

elsh committed:
Use real module names to load modules
Add ModuleInterface option meta tag for -module-alias
Add tests loading modules with -module-alias for swiftmodule and swiftinterface
with various loaders incl. serialized, explicit, and source loader.

elsh commented:
> Are module aliases intended to propagate through swift interface files as well?

Yup, added a test that verifies this.

elsh commented:
@swift-ci smoke test

adrian-prantl commented:
Could you add a simple test to test/DebugInfo that is like test/DebugInfo/Imports.swift, without the dwarfdump part, but with a  module alias, so we can document what the expected behavior in the debug info is going to be?

adrian-prantl commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -1218,13 +1224,14 @@ MemoryBufferSerializedModuleLoader::loadModule(SourceLoc importLoc,
     return nullptr;
 
   auto moduleID = path[0];
+  auto moduleName = Ctx.getRealModuleName(moduleID.Item).str();
```
Is the idea that the real module name is a long unique identifier (com.test.foo) and that path[0] is a short name (foo)?

Are module imports in serializes Swift modules using the long unique name?

MemoryBufferSerializedModuleLoader is only used by LLDB and with LLDB's global view of the application it may not generally be possible to unambiguously perform this translation. But if we only need to deal imports a user types into the expression evaluator then the right solution would be to intercept them in SwiftExpressionParser and rewrite them to the long unique name there, where we still have the contextual information.

elsh replied:
No it's not a long unique name. The module aliasing feature here allows a scenario like this: 
If you're building `MyModule`, and it has `import Foo`, but we want to alias `Foo` with another name under the hood due to a name collision, then we want to be able to run  `swift-frontend -module-name MyModule -module-alias Foo=Bar`, where the name `Bar` is the actual name of `Foo`, and there's Bar.swiftmodule on disk. The `getRealName()` returns `Bar`, while `getName()` returns `Foo`. If there's no `-module-alias` option used, `getRealName()` and `getName()` will have the same value.

adrian-prantl replied:
Each translation unit already stores all imported modules in the debug info. We could represent a module alias using a DIImportedEntity("short alias", DIModule("unique name")).
When the expression evaluator sees in import statement it checks the debug info for an alias and substituted the long name in the AST, so MemoryBufferSerializedModuleLoader only sees the long unique name.

If you agree that this would work, then we don't need to modify MemoryBufferSerializedModuleLoader in this patch.

elsh replied:
If a module a user typed in to import is not loaded yet, it won't work as the alias hasn't been stored, in which case they'll will have to import the physical module name explicitly until after it's loaded and the alias is stored. I'm not sure if the behavior seems consistent enough or might cause confusion.

elsh commented:
@swift-ci smoke test

elsh committed:
Update test

elsh committed:
Don't use getRealModuleName for lldb module loader for now

elsh commented:
@swift-ci smoke test

elsh committed:
Clean up a test

elsh commented:
@swift-ci smoke test

elsh committed:
Add tests with chained dependency scenarios with module aliasing

elsh commented:
@swift-ci smoke test

xedin reviewed (approved):
Looks good to me!

elsh committed:
Add tests with more chained deps scenarios

elsh committed:
Update tests due to windows escape char error

elsh commented:
@swift-ci smoke test
